### PR TITLE
2022 bms software

### PR DESCRIPTION
**Pull request by bdavis2020**

### CODE REVIEW

File: BMS_Software.h [hunk 1]
```diff
@@ -0,0 +1,187 @@
+// Battery Managment System (BMS) Software ////////////////////////////////////////////////////////////////////////////////////////////////////////////
+  // Main Header File
+//
+// Created for 2019 Valkyrie by: Jacob Lipina, jrlwd5
```
Comment (bdavis2020) [inline BMS_Software.h]: hmm

File: BMS_Software.h [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+#define BMS_Software
+
+
+//RoveComm setup
+#include "RoveComm.h"
+//RoveCommEthernet RoveComm; //extantiates a class
+
+// Pinmap ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Control Pins (output pins)
+#define BUZZER_CTR_PIN			37
```
Comment (bdavis2020) [inline BMS_Software.h]: consistent indentation

File: BMS_Software.h (hunk 1 above)
Comment (antrob25) [inline BMS_Software.h]: Change year and Author

File: BMS_Software.h (hunk 2 above)
Comment (antrob25) [inline BMS_Software.h]: Have even spacing for pin names

File: BMS_Software.h [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+//other system sensors
+#define PACK_V_MEAS_PIN			A10
+#define PACK_I_MEAS_PIN			A17
+#define TEMP_degC_MEAS_PIN		A12
+
+const int CELL_MEAS_PINS[] = {C1_V_MEAS_PIN, C2_V_MEAS_PIN, C3_V_MEAS_PIN, C4_V_MEAS_PIN, C5_V_MEAS_PIN, C6_V_MEAS_PIN, C7_V_MEAS_PIN, C8_V_MEAS_PIN};
+
+// Sensor Measurment: Constants and Calculations /////////////////////////////////////////////////////////////////////////////////////////////////////
+//
+// Tiva1294C RoveBoard Specs
```
Comment (bdavis2020) [inline BMS_Software.h]: teensy

File: BMS_Software.h [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  // Find at: https://www.pjrc.com/store/teensy41.html
+  // find at: https://www.pjrc.com/store/teensy41_pins.html
+#define SENSOR_SENSITIVITY   	0.0066 //V  3.3V / 1024 levels = 0.0032
+
+//value from past year
+#define SENSOR_SCALE         	0.5
+#define SENSOR_BIAS          	VCC * SENSOR_SCALE //Viout voltage when current is at 0A (aka quiescent output voltage)
+                          						 // Noise is 2mV, meaning the smallest current that the device is able to resolve is 0.3A
+// need to be tested for teensy
+#define CURRENT_MAX           	200000//mA; Current values must be sent over RoveComm as mA
```
Comment (bdavis2020) [inline BMS_Software.h]: Max current is 160000 mA

File: BMS_Software.h [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+#define CURRENT_MIN           	-196207//mA
+#define OVERCURRENT           100000 //mA
```
Comment (antrob25) [inline BMS_Software.h]: Change to 120

File: BMS_Software.h [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  // find at: https://www.pjrc.com/store/teensy41_pins.html
+#define SENSOR_SENSITIVITY   	0.0066 //V  3.3V / 1024 levels = 0.0032
+
+//value from past year
+#define SENSOR_SCALE         	0.5
+#define SENSOR_BIAS          	VCC * SENSOR_SCALE //Viout voltage when current is at 0A (aka quiescent output voltage)
+                          						 // Noise is 2mV, meaning the smallest current that the device is able to resolve is 0.3A
+// need to be tested for teensy
+#define CURRENT_MAX           	200000//mA; Current values must be sent over RoveComm as mA
+#define CURRENT_MIN           	-196207//mA
```
Comment (bdavis2020) [inline BMS_Software.h]: for 2% error, 156800 mA; probably need to test and play with this number

File: BMS_Software.h [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+#define OVERCURRENT           100000 //mA
+#define CURRENT_ADC_MIN     	0 //bits
+#define CURRENT_ADC_MAX     	4096 //bits
```
Comment (antrob25) [inline BMS_Software.h]: Comment should be 4096 values not bits as it is 12 bit ADC resolution

File: BMS_Software.h [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+// Voltage Measurments
+//cells
+#define CELL_VOLTS_MIN          2400//ADC value  //Lowest cell voltage I expect to be able to measure.
+#define CELL_VOLTS_MAX        4200 //ADC value //Highest possible cell voltage value we could measure.
+#define CELL_V_ADC_MIN        651 //from analogread
+#define CELL_V_ADC_MAX        922  //from analogread
+#define CELL_COUNT	8// number of batteries
```
Comment (bdavis2020) [inline BMS_Software.h]: number of modules

File: BMS_Software.h [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+#define CELL_UNDERVOLTAGE     2650 //mV //If any cell voltage reaches this voltage, turn off rover and BMS suicide.
```
Comment (bdavis2020) [inline BMS_Software.h]: 2650 to 2400

File: BMS_Software.h [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+// package related
+#define VOLTS_MIN               0 //mV    //Lowest possible pack voltage value we could measure.
+#define PACK_VOLTS_MAX        	33600 //mV	//Highest possible pack voltage value we could measure.
+#define PACK_UNDERVOLTAGE   	21600 //mV	//If pack voltage reaches this voltage, turn off rover and BMS suicide.
```
Comment (bdavis2020) [inline BMS_Software.h]: keep consistent with cell_undervoltage value

File: BMS_Software.h [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+#define PACK_LOWVOLTAGE     	25000 //mV	//If pack voltage reaches this voltage, notify of low voltage.
+#define PACK_SAFETY_LOW     	PACK_UNDERVOLTAGE - 4000 //mV	//Even though I shouldn't be need to measure a voltage lower than undervoltage
```
Comment (MaxR-mrdt) [inline BMS_Software.h]: not used, delete this

Comment (antrob25) [inline BMS_Software.h]: Remove as it is unused

File: BMS_Software.h (hunk 10 above)
Comment (MaxR-mrdt) [reply BMS_Software.h]: undervoltage = 19200 mV
lowvoltage = 21200

File: BMS_Software.h (hunk 11 above)
Comment (MaxR-mrdt) [reply BMS_Software.h]: same with pack_effective_zero and cell_effective_zero

File: BMS_Software.ino [hunk 12]
```diff
@@ -0,0 +1,1155 @@
+// Battery Managment System (BMS) Software System
+// Main cpp file (Needed to clarify that this is in c++)
```
Comment (antrob25) [inline BMS_Software.ino]: Remove Uneeded comments like c++ files and asking Anthony about Github

File: BMS_Software.ino [hunk 13]
```diff
@@ -0,0 +1,1155 @@
...
+rovecomm_packet packet;
+
+//timekeeping variables
+uint32_t last_update_time;
+
+//declare the Ethernet Server in the top level sketch with the requisite port ID any time you want to use RoveComm
+EthernetServer TCPServer(RC_ROVECOMM_BMSBOARD_PORT);
+
+
+//Setup and Main Loop
```
Comment (antrob25) [inline BMS_Software.ino]: Move global variables to the header file

File: BMS_Software.ino [hunk 14]
```diff
@@ -0,0 +1,1155 @@
...
+uint8_t error_report = 0;
+bool pinfault_state = false;
+int num_loop = 0; // battery temperature sensor state defualt to false in case of an error report
+bool sw_ind_state = false;
+uint32_t time_since_Rovecomm_update = 0; // intial time of Rovecomm updating
+uint32_t meas_batt_temp[NUM_TEMP_AVERAGE];
+float main_current; //PACK_I
+float cell_voltages[CELL_COUNT]; //Cell Voltages
+float pack_out_voltage; //PACK_V
+uint32_t batt_temp;   //TEMP_degC
```
Comment (antrob25) [inline BMS_Software.ino]: Change to float

File: BMS_Software.ino [hunk 15]
```diff
@@ -0,0 +1,1155 @@
...
+uint32_t last_update_time;
+
+//declare the Ethernet Server in the top level sketch with the requisite port ID any time you want to use RoveComm
+EthernetServer TCPServer(RC_ROVECOMM_BMSBOARD_PORT);
+
+
+//Setup and Main Loop
+uint8_t error_report = 0;
+bool pinfault_state = false;
+int num_loop = 0; // battery temperature sensor state defualt to false in case of an error report
```
Comment (antrob25) [inline BMS_Software.ino]: Specify int type for num_loop

File: BMS_Software.ino [hunk 16]
```diff
@@ -0,0 +1,1155 @@
...
+bool sw_ind_state = false;
+uint32_t time_since_Rovecomm_update = 0; // intial time of Rovecomm updating
```
Comment (ThisIsBrady) [inline BMS_Software.ino]: Never initialized before loop starts. Or use last_update_time

File: BMS_Software.ino [hunk 17]
```diff
@@ -0,0 +1,1155 @@
...
+  reactLowVoltage(cell_voltages); // Forgotten Logic Switch, and Over temperature.
+
+  getOutVoltage(pack_out_voltage);
+  reactEstopReleased();
+  reactForgottenLogicSwitch();
+
+  getBattTemp(batt_temp);
+  reactOverTemp();
+
+  if ((millis() - time_since_Rovecomm_update) >= ROVECOMM_UPDATE_DELAY)
```
Comment (antrob25) [inline BMS_Software.ino]: Change telemetry to interrupt based not polling

File: BMS_Software.ino [hunk 18]
```diff
@@ -0,0 +1,1155 @@
...
+    ////Serial.println(batt_temp);
+    RoveComm.write(RC_BMSBOARD_PACKI_MEAS_DATA_ID, RC_BMSBOARD_PACKI_MEAS_DATA_COUNT, main_current);
+    delay(100);
+    RoveComm.write(RC_BMSBOARD_PACKV_MEAS_DATA_ID, RC_BMSBOARD_PACKV_MEAS_DATA_COUNT, pack_out_voltage); // this if statement is created to write information of current, cell volt.,
+    delay(100);                                     // and battery temperature if the update delay is greater or equal to time since last update
+    RoveComm.write(RC_BMSBOARD_TEMP_MEAS_DATA_ID, RC_BMSBOARD_TEMP_MEAS_DATA_COUNT, batt_temp);
+    delay(100);
+    RoveComm.write(RC_BMSBOARD_CELLV_MEAS_DATA_ID,RC_BMSBOARD_CELLV_MEAS_DATA_COUNT, cell_voltages);
+
+    time_since_Rovecomm_update = millis();
```
Comment (antrob25) [inline BMS_Software.ino]: Use this last_update_time variable instead

File: BMS_Software.ino [hunk 19]
```diff
@@ -0,0 +1,1155 @@
+// Battery Managment System (BMS) Software System
```
Comment (antrob25) [inline BMS_Software.ino:1]: Change spacing to 4 spaces

File: BMS_Software.ino [hunk 20]
```diff
@@ -0,0 +1,1155 @@
...
+  ////Serial.println(analogRead(PACK_I_MEAS_PIN));
+
+  main_current =  map(analogRead(PACK_I_MEAS_PIN), CURRENT_ADC_MIN, CURRENT_ADC_MAX, CURRENT_MIN, CURRENT_MAX); //fetch pack current
+  // display of current functions
+  ////Serial.print("current:  ");
+  ////Serial.println(main_current);
+
+  if (main_current > OVERCURRENT)		//check current > overcurrent, if not return
+  {
+    delay(DEBOUNCE_DELAY); // debounce delay is to check twice in a short period of time
```
Comment (antrob25) [inline BMS_Software.ino]: There might be a Teensy Library for debouncing to correctly debounce the reading

File: BMS_Software.ino [hunk 21]
```diff
@@ -0,0 +1,1155 @@
...
+  }
+  if (adc_reading < TEMP_ADC_MIN)
+  {
+    adc_reading = TEMP_ADC_MIN;
+  }
+
+  meas_batt_temp[num_meas_batt_temp] = (1060 * (map(adc_reading, TEMP_ADC_MIN, TEMP_ADC_MAX, TEMP_MIN, TEMP_MAX)) / 1000); // function of measuring battery temp with
+  num_meas_batt_temp++;                                                                                                    // increasing interval of the function
+
+  batt_temp = (925 * (map(analogRead(TEMP_degC_MEAS_PIN), TEMP_ADC_MIN, TEMP_ADC_MAX, TEMP_MIN, TEMP_MAX)) / 1000); // mapping analog signals of battery temp data
```
Comment (ThisIsBrady) [inline BMS_Software.ino]: What are the 1060 and 925 here? Maybe put them as consts

File: BMS_Software.ino [hunk 22]
```diff
@@ -0,0 +1,1155 @@
...
+        if (adc_reading < CELL_V_ADC_MIN) 	//handle if reading between 0V and 2.4V: lower than expecting power
+        {
+          adc_reading = CELL_V_ADC_MIN; 			 
+        }     
+        else if (adc_reading > CELL_V_ADC_MAX) 	//handle if reading above 3.3V	:above expecting voltage
+        {
+          adc_reading = CELL_V_ADC_MAX;
+        }
+        cell_voltages[i] = ((map(adc_reading, CELL_V_ADC_MIN, CELL_V_ADC_MAX, CELL_VOLTS_MIN, CELL_VOLTS_MAX)));                                                                                                                                                                                                                               // end if
+        if ((map(adc_reading, CELL_V_ADC_MIN, CELL_V_ADC_MAX, CELL_VOLTS_MIN, CELL_VOLTS_MAX) <= CELL_UNDERVOLTAGE)
```
Comment (antrob25) [inline BMS_Software.ino]: Map function should be replaced to cell_voltages[i].

File: BMS_Software.ino [hunk 23]
```diff
@@ -0,0 +1,1155 @@
...
+          cell_undervoltage_count++;
+        } // end if
+      }     // end if
+  }         // end for
+  return;
+} // end func
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void getOutVoltage(float &pack_out_voltage)
```
Comment (antrob25) [inline BMS_Software.ino]: Change to getPackVoltage

File: BMS_Software.ino [hunk 24]
```diff
@@ -0,0 +1,1155 @@
...
+{
+  int adc_reading = analogRead(PACK_V_MEAS_PIN);
+  ////Serial.print("adc vout : ");
+  ////Serial.println(adc_reading);
+  ////////////////////////////////////HERE////////////////////////////////////////////////
+  pack_out_voltage = ((1269 * map(adc_reading, PACK_V_ADC_MIN, PACK_V_ADC_MAX, VOLTS_MIN, PACK_VOLTS_MAX)) / 1000); // 1269 is the voltage to analog signal ratio
```
Comment (antrob25) [inline BMS_Software.ino]: Make 1269 into a constant called VoltageToSignalRatio

File: BMS_Software.ino [hunk 25]
```diff
@@ -0,0 +1,1155 @@
...
+  ////Serial.println(pack_out_voltage);
+  if (pack_out_voltage < PACK_UNDERVOLTAGE)
+  {
+    delay(DEBOUNCE_DELAY);
+	  adc_reading = analogRead(PACK_V_MEAS_PIN);
+	  pack_out_voltage = ((1269 * map(adc_reading, PACK_V_ADC_MIN, PACK_V_ADC_MAX, VOLTS_MIN, PACK_VOLTS_MAX)) / 1000); // previously1369
+	  if(pack_out_voltage < PACK_UNDERVOLTAGE)
+    {
+      //raise a error flag for RoveComm pack under voltage
+      pack_undervoltage_state = true;
```
Comment (antrob25) [inline BMS_Software.ino]: reset bool

File: BMS_Software.ino [hunk 26]
```diff
@@ -0,0 +1,1155 @@
...
+  if (adc_reading > TEMP_ADC_MAX)
+  {
+    adc_reading = TEMP_ADC_MAX; // analog pin rewrite with new parameters of given if statements from adc
+  }
+  if (adc_reading < TEMP_ADC_MIN)
+  {
+    adc_reading = TEMP_ADC_MIN;
+  }
+
+  meas_batt_temp[num_meas_batt_temp] = (1060 * (map(adc_reading, TEMP_ADC_MIN, TEMP_ADC_MAX, TEMP_MIN, TEMP_MAX)) / 1000); // function of measuring battery temp with
```
Comment (antrob25) [inline BMS_Software.ino]: Make random numbers to into constants

File: BMS_Software.ino [hunk 27]
```diff
@@ -0,0 +1,1155 @@
...
+  return;
+} // end func
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void updateLCD(int32_t batt_temp, float cell_voltages[])
+{
+  // Serial.print("**************************************************************************");
+
+  const int NUM_CELLS = 8; // this should be in header?
```
Comment (ThisIsBrady) [inline BMS_Software.ino]: Defined in header as CELL_COUNT

File: BMS_Software.ino [hunk 28]
```diff
@@ -0,0 +1,1155 @@
...
+{
+  ////Serial.println("reactLowVoltage");
+  if ((cell_voltage[0] > PACK_UNDERVOLTAGE) && (cell_voltage[0] <= PACK_LOWVOLTAGE) && (low_voltage_state = false)) // first instance of low voltage
+  {
+    low_voltage_state = true;
+    notifyLowVoltage();
+    time_of_low_voltage = millis();
+    num_low_voltage_reminder = 1;
+  }                                                                                                                     // end if
+  else if ((cell_voltage[0] > PACK_UNDERVOLTAGE) && (cell_voltage[0] <= PACK_LOWVOLTAGE) && (low_voltage_state = true)) // following instances of low voltage
```
Comment (ThisIsBrady) [inline BMS_Software.ino]: third condition needs == not =, or remove and use low_voltage_state

File: BMS_Software.ino [hunk 29]
```diff
@@ -0,0 +1,1155 @@
...
+    estop_released_beep = true;
+    notifyEstopReleased();
+  }
+}
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void reactLowVoltage(float cell_voltage[CELL_COUNT])
+{
+  ////Serial.println("reactLowVoltage");
+  if ((cell_voltage[0] > PACK_UNDERVOLTAGE) && (cell_voltage[0] <= PACK_LOWVOLTAGE) && (low_voltage_state = false)) // first instance of low voltage
```
Comment (ThisIsBrady) [inline BMS_Software.ino]: third condition needs == not =, or remove and use !low_voltage_state

File: BMS_Software.ino [hunk 30]
```diff
@@ -0,0 +1,1155 @@
...
+void setEstop(uint8_t data)
+{
+  if (data == 0)
+  {
+    digitalWrite(PACK_GATE_CTR_PIN, LOW);
+
+    notifyEstop();
+    delay(100);
+
+    digitalWrite(PACK_GATE_CTR_PIN, HIGH); // BMS Suicide
```
Comment (antrob25) [inline BMS_Software.ino]: Remove the reenable

**Review by antrob25 — CHANGES_REQUESTED** [review]
Nice job ;smirk;

File: BMS_Software.ino [hunk 31]
```diff
@@ -0,0 +1,1148 @@
...
+          Serial3.write(0x48);//H
+          Serial3.write(0x41);//A
+          Serial3.write(0x52);//R
+          Serial3.write(0x44);//D*/
+    }
+    Serial3.write('|'); // Setting character
+    Serial3.write('-'); // Clear display
+}
+
+void asterisks()
```
Comment (antrob25) [inline BMS_Software.ino]: Change Valkyrie to Prometheus

**Review by antrob25 — CHANGES_REQUESTED** [review]

**Pull request merged**